keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,479 @@
+- In case an administrator enables this feature and the webhook kubeconfig file is invalid or doesn't exist at the specified path, `kube-apiserver` on that node will not be able to start. This can be mitigated by fixing the malformed values.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We would like to introduce a structured file format which allows authorization to be configured using a flag (`--authorization-config-file`) which accepts a path to a file on the disk. This feature can be enabled or disabled by the explicit feature flag `--authorization-config-from-file`. The proposed structure is illustrated below:
+```
+apiVersion: apiserver.config.k8s.io/v1
+kind: AuthorizationConfiguration
+authorizers:
+- type: Webhook
+  webhook:
+    kubeConfigFile: <path-to-deny-webhook-config>
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: Deny
+- type: Node
+- type: RBAC
+- type: Webhook
+  webhook:
+    kubeConfigFile: <path-to-allow-webhook-config>
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: NoOpinion
+```
+Validation will allow multiple authorizers of type "Webhook" to be added to the config, but one authorizer each for other types. The ordering of this chain will be decided by the order specified in the file.
+
+The keys `kubeConfigFile`, `authorizedTTL`, `unauthorizedTTL` and `subjectAccessReviewVersion` version accept values corresponding to flags `--authorization-webhook-config-file`, `--authorization-webhook-cache-authorized-ttl`, `--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version` respectively.
+
+This diagram shows how the configuration flags map to configuration file keys for a webhook:
+![](https://i.imgur.com/ZEuqcKe.jpg)
+
+The `onError` will be a required field which allow users to specify whether or not request is denied if the webhook errors out or is unreachable. This allows `SubjectAccessReviews` to be more definitive.
+
+Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding flag is not supplied. While configuring authorization modes using the file config, the version supported by a webhook has to be mentioned using a required field `subjectAccessReviewVersion`.
+
+Authentication of webhooks Webhooks will continue to be configured by passing a path to a file resembling a `kubeconfig`. Eventually, we can make a decision about moving webhook configuration to the proposed authorization configuration.
+
+The config file would accept a path to the policy file if the auth mode type is ABAC.
```

Comment:
why would we choose to allow ABAC?  If we don't want to encourage it, it seems more straightforward to not support it in this new file.

Reply:
We weren't sure if ABAC was going to be deprecated. In that case, we can remove its support from the config file.

Reply:
Since we discussed to not deprecate any old features while implementing the Structured Configuration, going to keep support ABAC in the configuration format.